presto-main/src/main/java/com/facebook/presto/sql/MaterializedViewUtils.java

```diff
@@ -0,0 +1,94 @@
+     * In case of 1 materialized view defined on N base tables via join, union, etc, this method helps compute
+     * all the N base table columns that one materialized view column maps to.
+     * It calls on {@link MaterializedViewColumnMappingExtractor} to get all base table columns mapped by join, union, etc,
+     * and then uses them to expand the 1-to-1 column mapping derived from {@code analysis} to the 1-to-N column mapping.
+     * <p>
+     * For example, given SELECT column_a AS column_x FROM table_a JOIN table_b ON (table_a.column_a = table_b.column_b),
+     * the 1-to-1 column mapping from {@code analysis} is column_x -> table_a.column_a. Mapped base table columns are
+     * [table_a.column_a, table_b.column_b]. Then it will return a 1-to-N column mapping column_x -> {table_a.column_a, table_b.column_b}.
+     */
+    public static Map<String, Map<SchemaTableName, String>> computeMaterializedViewToBaseTableColumnMappings(Analysis analysis)
+    {
+        checkState(
+                analysis.getStatement() instanceof CreateMaterializedView,
+                "Only support the computation of column mappings when analyzing CreateMaterializedView");
+
+        ImmutableMap.Builder<String, Map<SchemaTableName, String>> fullColumnMapping = ImmutableMap.builder();
+
+        Query viewQuery = ((CreateMaterializedView) analysis.getStatement()).getQuery();
+        Map<String, TableColumn> originalColumnMapping = getOriginalColumnsFromAnalysis(viewQuery, analysis);
+
+        List<List<TableColumn>> mappedBaseColumns = MaterializedViewColumnMappingExtractor.extractMappedBaseColumns(analysis);
+
+        for (String viewColumn : originalColumnMapping.keySet()) {
+            TableColumn originalBaseColumn = originalColumnMapping.get(viewColumn);
+
+            Map<SchemaTableName, String> fullBaseColumns = new HashMap<>();
+
+            mappedBaseColumns.forEach(mappedBaseColumnPair -> {
+                if (originalBaseColumn.equals(mappedBaseColumnPair.get(0))) {
+                    fullBaseColumns.put(mappedBaseColumnPair.get(1).getTableName(), mappedBaseColumnPair.get(1).getColumnName());
+                }
+                else if (originalBaseColumn.equals(mappedBaseColumnPair.get(1))) {
+                    fullBaseColumns.put(mappedBaseColumnPair.get(0).getTableName(), mappedBaseColumnPair.get(0).getColumnName());
+                }
+            });
+
+            fullBaseColumns.put(originalBaseColumn.getTableName(), originalBaseColumn.getColumnName());
+
+            fullColumnMapping.put(viewColumn, ImmutableMap.copyOf(fullBaseColumns));
+        }
+
+        return fullColumnMapping.build();
+    }
+
+    private static Map<String, TableColumn> getOriginalColumnsFromAnalysis(Node viewQuery, Analysis analysis)
+    {
+        return analysis.getOutputDescriptor(viewQuery).getVisibleFields().stream()
+                .filter(field -> field.getOriginTable().isPresent() && field.getOriginColumnName().isPresent())
```

Comment:
Good question, in the case of expression or aggregation applied to a column in the select list, the computed column mappings will not include info for the column. We have no way to get the mapping for the column (and do not have strong need now), since the final MV column could possibly derive from more than one column from the base tables.

Reply:
Would be good to add a comment about this limitation.

Reply:
As you can see from the impl, that is kind of the design of Field.getOriginTable() and Field.getOriginColumnName() in general, and not particular to MV.